File: backend/api/schemas/forward_testing.py
"""
Forward Testing API schemas — Phase 4C.5.

Request/response models for:
    POST   /forward-tests
    GET    /forward-tests
    GET    /forward-tests/{session_id}
    POST   /forward-tests/{session_id}/run-cycle
    POST   /forward-tests/{session_id}/pause
    POST   /forward-tests/{session_id}/resume
    POST   /forward-tests/{session_id}/terminate
    GET    /forward-tests/{session_id}/signals
    GET    /forward-tests/{session_id}/bars

Security invariants (enforced at this layer):
    - No file_path in any response
    - No credential secrets (credential_id metadata is safe; raw key is never surfaced)
    - No strategy_json in list or summary responses
    - user_id never in list responses (bulk PII reduction)
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class CreateForwardTestSessionRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    draft_id: str
    symbol: str
    timeframe: str
    source_mode: Literal["provider", "catalog"]
    provider_name: str | None = None   # required when source_mode="provider"
    catalog_id: str | None = None      # required when source_mode="catalog"
    credential_id: str | None = None   # vault credential_id for polygon; never raw key
    exchange: str = "NASDAQ"           # needed to build Instrument / DatasetIdentity
    asset_class: str = "equity"        # needed to build Instrument / DatasetIdentity


# ---------------------------------------------------------------------------
# Strategy snapshot summary (no strategy_json — never expose in responses)
# ---------------------------------------------------------------------------

class ForwardTestStrategySnapshotResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    draft_id: str
    display_name: str
    lifecycle_status: str
    snapshot_hash: str
    # description and semantics_hash omitted; captured_at omitted from summary


# ---------------------------------------------------------------------------
# Session summary — lightweight list item
# ---------------------------------------------------------------------------

class ForwardTestSessionSummaryResponse(BaseModel):
    """
    Lightweight session record for list views.

    No strategy_json.  No user_id.  No file_path.  No secrets.
    """
    model_config = ConfigDict(frozen=True)

    session_id: str
    status: str
    symbol: str
    timeframe: str
    source_mode: str
    provider_name: str | None
    catalog_id: str | None
    created_at: str
    updated_at: str
    last_processed_bar_timestamp: str | None
    bars_evaluated: int
    signals_recorded: int
    strategy_snapshot: ForwardTestStrategySnapshotResponse


# ---------------------------------------------------------------------------
# Session detail — full metadata, still no strategy_json or user_id
# ---------------------------------------------------------------------------

class ForwardTestSessionDetailResponse(ForwardTestSessionSummaryResponse):
    """
    Full session detail for GET /forward-tests/{session_id}.

    Extends the summary with lifecycle and evaluation metadata.
    """
    draft_id: str
    lifecycle_status_at_activation: str
    warmup_bars_required: int
    warmup_bars_processed: int
    signal_eligible_bars_processed: int
    activation_timestamp: str | None
    failure_reason: str | None
    error_category: str | None
    exchange: str
    asset_class: str


# ---------------------------------------------------------------------------
# Cycle result — returned by run-cycle
# ---------------------------------------------------------------------------

class ForwardTestCycleResultResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    session_id: str
    status: str
    bars_fetched: int
    bars_processed: int
    warmup_bars_processed: int
    signal_eligible_bars_processed: int
    signals_generated: int
    signals_suppressed: int
    last_processed_bar_timestamp: str | None
    gap_detected: bool
    provider_failure: bool
    activated: bool
    message: str | None


# ---------------------------------------------------------------------------
# Signal response — immutable record, no secrets
# ---------------------------------------------------------------------------

class ForwardTestSignalResponse(BaseModel):
    """
    Safe signal record for API responses.

    No file_path.  No credential secrets.  No raw provider error messages.
    No fill, position, equity, or P&L fields.
    """
    model_config = ConfigDict(frozen=True)

    signal_id: str
    session_id: str
    bar_timestamp: str
    signal_timestamp: str
    signal_direction: str
    rule_id: str
    bar_open: float
    bar_high: float
    bar_low: float
    bar_close: float
    bar_volume: float
    feature_values_at_signal: dict[str, float]
    warmup_satisfied: bool
    strategy_snapshot_hash: str
    symbol: str
    timeframe: str
    provider_name: str | None
    catalog_id: str | None
    created_at: str


# ---------------------------------------------------------------------------
# Bar response — optional endpoint, safe subset
# ---------------------------------------------------------------------------

class ForwardTestBarResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    bar_index: int
    bar_timestamp: str
    open: float
    high: float
    low: float
    close: float
    volume: float
    is_warmup_bar: bool
    processed_at: str

File: frontend/src/types/forwardTesting.ts
/**
 * Forward Testing TypeScript types — Phase 4C.5.
 *
 * Mirror of backend/api/schemas/forward_testing.py.
 * No strategy_json exposed. No user_id in list/summary types.
 * No file_path. No credential secrets.
 */

export type ForwardTestSessionStatus =
  | 'pending'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'terminated'

export interface ForwardTestStrategySnapshot {
  draft_id: string
  display_name: string
  lifecycle_status: string
  snapshot_hash: string
}

export interface ForwardTestSessionSummary {
  session_id: string
  status: ForwardTestSessionStatus
  symbol: string
  timeframe: string
  source_mode: 'provider' | 'catalog'
  provider_name: string | null
  catalog_id: string | null
  created_at: string
  updated_at: string
  last_processed_bar_timestamp: string | null
  bars_evaluated: number
  signal_eligible_bars_processed: number
  signals_recorded: number
  strategy_snapshot: ForwardTestStrategySnapshot
}

export interface ForwardTestSessionDetail extends ForwardTestSessionSummary {
  draft_id: string
  lifecycle_status_at_activation: string
  warmup_bars_required: number
  warmup_bars_processed: number
  activation_timestamp: string | null
  failure_reason: string | null
  error_category: string | null
  exchange: string
  asset_class: string
  /** Set by the FT-2 scheduler; null until the scheduler has run at least once. */
  last_cycle_attempted_at: string | null
}

export interface ForwardTestCycleResult {
  session_id: string
  status: ForwardTestSessionStatus
  bars_fetched: number
  bars_processed: number
  warmup_bars_processed: number
  signal_eligible_bars_processed: number
  signals_generated: number
  signals_suppressed: number
  last_processed_bar_timestamp: string | null
  gap_detected: boolean
  provider_failure: boolean
  activated: boolean
  message: string | null
}

export interface ForwardTestSignal {
  signal_id: string
  session_id: string
  bar_timestamp: string
  signal_timestamp: string
  signal_direction: 'entry_long' | 'entry_short' | 'exit_long' | 'exit_short' | 'no_action'
  rule_id: string
  bar_open: number
  bar_high: number
  bar_low: number
  bar_close: number
  bar_volume: number
  feature_values_at_signal: Record<string, number>
  warmup_satisfied: boolean
  strategy_snapshot_hash: string
  symbol: string
  timeframe: string
  provider_name: string | null
  catalog_id: string | null
  created_at: string
  /** Timestamp of the next bar after signal bar; null when signal fires on the final bar in the cycle. */
  actionable_from_bar_timestamp: string | null
}

export interface ForwardTestBar {
  bar_index: number
  bar_timestamp: string
  open: number
  high: number
  low: number
  close: number
  volume: number
  is_warmup_bar: boolean
  processed_at: string
}

export interface CreateForwardTestSessionRequest {
  draft_id: string
  symbol: string
  timeframe: string
  source_mode: 'provider' | 'catalog'
  provider_name?: string | null
  catalog_id?: string | null
  credential_id?: string | null
  exchange?: string
  asset_class?: string
}

/** Context extracted from a backtest report to prefill forward-test session creation. */
export interface ForwardTestPrefill {
  draft_id:       string
  draft_name?:    string
  symbol?:        string
  timeframe?:     string
  provider_name?: string
  exchange?:      string
  asset_class?:   string
}
